compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1479,9 +1479,39 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
     /** Like tp1 <:< tp2, but returns false immediately if we know that
      *  the case was covered previously during subtyping.
+     *
+     *  A type has been covered previously in subtype checking if it
+     *  is some combination of TypeRefs that point to classes, where the
+     *  combiners are AppliedTypes, RefinedTypes, RecTypes, And/Or-Types or AnnotatedTypes.
+     *
+     *  The exception is that if both sides contain OrTypes, the check hasn't been covered.
+     *  See #17465.
      */
     def isNewSubType(tp1: Type): Boolean =
-      if (isCovered(tp1) && isCovered(tp2))
+
+      def isCovered(tp: Type): (Boolean, Boolean) =
+        var containsOr: Boolean = false
+        @annotation.tailrec def recur(todos: List[Type]): Boolean = todos match
+          case tp :: todos =>
+            tp.dealiasKeepRefiningAnnots.stripTypeVar match
+              case tp: TypeRef =>
+                if tp.symbol.isClass && tp.symbol != NothingClass && tp.symbol != NullClass then recur(todos)
+                else false
+              case tp: AppliedType => recur(tp.tycon :: todos)
```

Comment:
These nested recurrences might show up to be really expensive. Why do we need them?

Reply:
The recursion is equivalent to the one before. I refactored it to take a list of types, which is essentially a working list of recursive invocations, to make the function tail-recursive. I thought that this saves the stack and could lead to better performance, but I could revert this refactorization to avoid passing around a list during recursion and creating a new list object at each new invocation.

Reply:
Previously we did not recurse in arguments of `AppliedTypes`, just in the type constructor. That's the one I would expect to matter most. 

Generally, I think using stack is cheaper than allocating.

Reply:
Here we are not recursing on the arguments of the `AppliedType`: note that the `todos` is the working list, not the arguments of the `AppliedType`.

Thanks for pointing this out! I will revert this rewriting, and in the future I'll prefer reducing heap allocations over achieving tail recursion.

Reply:
Ah, right, I had misread that!

Reply:
In fact, maybe it's simplest to just use `Int`s as the result of `recur`. Then `&&` could be replaced by `min`.

Reply:
Yes, I just pushed the [commit](https://github.com/lampepfl/dotty/pull/18213/commits/380bd1b91a6ca0edbf4a555ea48455bdd221c5a5) that reverts the tailrec rewriting and uses an integer-based representation for the result of `isCovered`, interpreted bit-wisely. `min` is used to aggregate results as you suggested. (A bitwise `&` was first used for aggregation but later I found that `min` is more logical). We use bitwise operations to analyze the results as well.